### DISCUSSION TITLE
Godot 4 Firebase Integration on HTML5 Export

### DISCUSSION THREAD

**Discussion by Piesrgr8** [GodotNuts/GodotFirebase#396] [Q&A]
Hello there! I am using this add-on as part of a capstone project. We thought it would be really cool to use Firebase as a database for logging the fastest times to complete an online escape room. But, while the Firebase connectivity works on local, it doesn't execute any statements in our script on HTML5. Currently, our export does contain .env and .txt as part of the game. 
![Screenshot_20240427_180008](https://github.com/GodotNuts/GodotFirebase/assets/11251881/81f783ef-5861-47dc-b141-f9cd18f64cbf)

Code for posting:
```
func make_post_request(scenario, time, username, finished):
	var collection: FirestoreCollection = Firebase.Firestore.collection("scores")
	var query: FirestoreQuery = FirestoreQuery.new()
	query.from("scores")
	query.get("username")
	var query_task: FirestoreTask = Firebase.Firestore.query(query)
	var result: Array = await Firebase.Firestore.query(query).result_query
	var latest_id = 1
	if result.size() > 0:
		for score in result:
			if score != null and score['doc_fields']['username'].contains(username):
				latest_id += 1
		collection.add(username + "_" + str(latest_id), {
			"scenario": scenario,
			"time": time,
			"username": username + "_" + str(latest_id),
			"finished": finished,
			"createdAt": Time.get_datetime_string_from_system()
		})
	else:
		collection.add(username + "_" + str(latest_id), {
		"scenario": scenario,
		"time": time,
		"username": username + "_" + str(latest_id),
		"finished": finished,
		"createdAt": Time.get_datetime_string_from_system()
	})
```

Code for getting:
```
func make_get_request(scenario_num) -> String:
	var query: FirestoreQuery = FirestoreQuery.new()
	query.from("scores")
	query.where("finished", FirestoreQuery.OPERATOR.EQUAL, 1)
	query.order_by("time", FirestoreQuery.DIRECTION.ASCENDING)
	query.limit(5)
	var query_task: FirestoreTask = Firebase.Firestore.query(query)
	var result: Array = await Firebase.Firestore.query(query).result_query
	var scores_dict: Dictionary = {}
	for score in result:
		if score.get("doc_fields").get("scenario") == scenario_num:
			print(score)
			var username = score['doc_fields']['username']
			var time = score['doc_fields']['time']
			var scenario = score['doc_fields']['scenario']
			if scores_dict.has(username):
				scores_dict[username].append({"time": time, "scenario": scenario})
			else:
				scores_dict[username] = [{"time": time, "scenario": scenario}]
	return JSON.stringify(scores_dict)
```

**Comment by WolfgangSenff** [comment]
It does work if things are set up properly. Go download this, put in your own .env file and test Firestore for web. It should work I think. Can't test now myself: https://github.com/GodotNuts/GodotFirebaseTestHarness

**Comment by WolfgangSenff** [comment]
Incidentally, which call is failing? The error looks very weird now that I can know at it closer. Still on my phone, might not have time to get to it tonight.

**Reply by Piesrgr8** [reply]
Currently, the function that is failing in the error is the post call.

**Reply by WolfgangSenff** [reply]
I think there's something else going on here. When I ran the test harness on the web and ran the Firestore tests, everything works fine. While that's not the be-all end-all, it's pretty indicative that things are working the way we expect. You're getting a number of errors that look like they're occurring (although not sure) because you're not awaiting the task that comes back from `collection.add()`.

	var add_task : FirestoreTask = _collection.add("Document1", {'name': 'Document1', 'active': 'true'})
	_document = await add_task.add_document

That's what it looks like in our test harness. Also, where do you do the call to make_post_request? I'll need to see more context to have a better idea of how this is breaking. The header thing might be not having authentication setup, although I would think the anonymous auth, which looks like it's kicking in, would be handling it okay.

**Reply by Piesrgr8** [reply]
This function would get called once the user has finished the escape room or quit in the middle of it.
```
func send_data(completed):
	finished_time = calculate_time($HUD/Timer/Label.text)
	$DBoperations.make_post_request(scenario_num, finished_time, username, completed)


func get_data():
	$DBoperations.make_get_request()
```
completed would be 1 or 0 to be treated as a boolean (just because we had a recent switch from our php db to firebase). DBoperations is a scene with the post script attached to it. Let me know if there is anything I missed here!

**Reply by WolfgangSenff** [reply]
Alright, let's try to clear this up a little bit more. We're gonna do some old fashioned debugging. Put print statements on every line in make_post_request with numbers on them to see where you get to before the error occurs and then let me know. Normally this is the type of thing where I'd send you a contract and get paid to solve this, but since I don't recall you ever asking any questions before (?), the first complex one is always free! But if you want me to look at the project, I can, and you can add me on github or wherever, or just zip it and send it. I say github because it's easy to give me access and remove access. Either way, try the print statements and let me know where you get to specifically.

**Reply by WolfgangSenff** [reply]
What it honestly looks like is happening is that it seems like you're sending two requests probably in very quick succession somewhere, as if you're sending the data in _process or something. But that is probably not it. Let me know if you're calling those two functions quickly in succession anywhere, or the same one multiple times. We have queuing for them so they should automatically queue up, but I can't see why yours would fail, given that.

**Reply by Piesrgr8** [reply]
Is there anything we would need to do to clear the http-header being null? Do we need to modify our rules for the firestore db? As instructed, I awaited the calls for adding the documents and queries, but that hasn't fixed the issue yet.

**Reply by WolfgangSenff** [reply]
That's one possibility, though I don't think it's that. My best guess is just some kind of problem with the authentication. Are you logged in when you do this?

**Reply by Piesrgr8** [reply]
From the script, I am logged in anonymously.

**Reply by WolfgangSenff** [reply]
That's doing it automatically. I wonder if you just wait until you're logged in and call to login anonymously by hand, what would happen? Did you try my test harness?

**Reply by Piesrgr8** [reply]
I started to let the authentication happen once the game starts, but that still lead to the same issue. Also, the test harness is currently in Godot 3, while the game and client I am working on is Godot 4. Let me download godot 3 and see whats happening.

**Reply by WolfgangSenff** [reply]
Get the 4.x branch.

**Reply by WolfgangSenff** [reply]
You're probably authenticating and calling to get the data in a singleton, aren't you? I wish I didn't have to guess. That's what I suspect, and authentication isn't finishing before it sends the request, so it tries to reauthenticate while in the middle of a call to authenticate. That alone could be causing that header issue, and putting your authentication into a weird state.

**Reply by Piesrgr8** [reply]
My bad 😅 Okay, I tested the Firestore part of the addon, adding my env and the user and password it tests with gave me this error.
![Screenshot 2024-04-28 195920](https://github.com/GodotNuts/GodotFirebase/assets/11251881/423be6b7-5d31-4d28-802b-448ecf26cc1d)
It didn't reach the 2nd test.

**Reply by WolfgangSenff** [reply]
That's weird. I don't think delete is the first test. Maybe I accidentally am doing some cleanup without checking. One second, gonna take a minute to look this up.

**Reply by WolfgangSenff** [reply]
Alright, go into this file and comment out this line. https://github.com/GodotNuts/GodotFirebaseTestHarness/blob/d9414f5fafbd9d6ccd395d469db566c72dc83db7/tests/firestore/firestore.gd#L82

**Reply by Piesrgr8** [reply]
Commented out, passed up until delete document, outputting this error:
![Screenshot 2024-04-28 201105](https://github.com/GodotNuts/GodotFirebase/assets/11251881/0f6db26d-8453-469a-90d4-ddbbc2f25ba3)

**Reply by WolfgangSenff** [reply]
Okay, comment out that darn test. 🤣 Looks like I forgot to update the test after changing delete!

**Reply by Piesrgr8** [reply]
Got it, Firestore tests passed, with the exception of the indexing on the query. Testing auth, we have this from the console:
STARTING AUTH TESTS

Trying to signup...

Trying to login...
  error code: 400
  message: INVALID_LOGIN_CREDENTIALS
  There has been a failure
  
Debugger shows no errors.

**Reply by WolfgangSenff** [reply]
That's normal - we require having a specific address allowed in the auth database. So it looks like it is indeed with your setup then. I recommend trying to wait to make any requests until you're fully logged in, then let me know if anything weird happens still. If the Firestore tests run, you should be good.

**Reply by Piesrgr8** [reply]
I guess the only difference here is that the authentication is done as an anonymous user. I'll try and just define a new user and have godot connect with that instead.

**Reply by WolfgangSenff** [reply]
Anonymous is fine, you just have to let it finish before making any requests.
